### PR TITLE
Value map combobox filter

### PR DESCRIPTION
**Pull request by domi4484**
Make value map widget searchable

https://github.com/opengisch/QField/assets/9881900/c9e019ef-ea1a-42c1-b5a9-8cede40dd47b

### CODE REVIEW

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 520f4189fcc052cf71ba647a73266d685598bcee: [**arm64-android**](https://qfieldapks.sos-ch-dk-2.exoscale-cdn.com/ci-builds/qfield-dev-5192-4a563f9408dd2f5a1553a95af08eebc4c692c3db-android-arm64.apk)
<details>
<summary>Other architectures</summary>

- [arm-neon-android](https://qfieldapks.sos-ch-dk-2.exoscale-cdn.com/ci-builds/qfield-dev-5192-4a563f9408dd2f5a1553a95af08eebc4c692c3db-android-arm-neon.apk)
- [x64-android](https://qfieldapks.sos-ch-dk-2.exoscale-cdn.com/ci-builds/qfield-dev-5192-4a563f9408dd2f5a1553a95af08eebc4c692c3db-android-x64.apk)
- [x86-android](https://qfieldapks.sos-ch-dk-2.exoscale-cdn.com/ci-builds/qfield-dev-5192-4a563f9408dd2f5a1553a95af08eebc4c692c3db-android-x86.apk)
- [**all access arm64-android**](https://qfieldapks.sos-ch-dk-2.exoscale-cdn.com/ci-builds/qfield_all_access-dev-5192-4a563f9408dd2f5a1553a95af08eebc4c692c3db-android-arm64.apk)
- [all access arm-neon-android](https://qfieldapks.sos-ch-dk-2.exoscale-cdn.com/ci-builds/qfield_all_access-dev-5192-4a563f9408dd2f5a1553a95af08eebc4c692c3db-android-arm-neon.apk)
- [all access x64-android](https://qfieldapks.sos-ch-dk-2.exoscale-cdn.com/ci-builds/qfield_all_access-dev-5192-4a563f9408dd2f5a1553a95af08eebc4c692c3db-android-x64.apk)
- [all access x86-android](https://qfieldapks.sos-ch-dk-2.exoscale-cdn.com/ci-builds/qfield_all_access-dev-5192-4a563f9408dd2f5a1553a95af08eebc4c692c3db-android-x86.apk)
</details>

File: src/qml/RelationCombobox.qml [hunk 1]
```diff
@@ -114,7 +114,7 @@ Item {
             featureListModel.searchTerm = searchField.displayText
           }
 
-          Keys.onPressed: {
+          Keys.onPressed: (event)=> {
```
Comment (nirvn) [inline src/qml/RelationCombobox.qml:117]: @domi4484 , thanks for hunting these, the more we can fix now, the better we'll be when Qt decided to kill injected parameters altogether (which I hope will be sooner than later! :) )

File: src/qml/editorwidgets/ValueMap.qml [hunk 2]
```diff
@@ -77,28 +84,235 @@ EditorWidgetBase {
...
 
-      Rectangle {
         visible: enabled
+
+        onClicked: {
+            searchFeaturePopup.open()
+        }
+    }
+
+    Popup {
```
Comment (nirvn) [inline src/qml/editorwidgets/ValueMap.qml:135]: There's a lot of code duplication between that and the relation combobox. Could we try and generalize that into a reusable item for both the RelationCombobox and this value map editor widget?

File: src/core/valuemapmodelbase.cpp [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgsvaluemapfieldformatter.h"
+#include "valuemapmodel.h"
+#include "valuemapmodelbase.h"
+
+#include <QDebug>
```
Comment (nirvn) [inline src/core/valuemapmodelbase.cpp:23]: This should go away.

File: src/core/valuemapmodel.h [hunk 4]
```diff
@@ -50,26 +46,20 @@ class ValueMapModel : public QAbstractListModel
...
 
-    /**
-     * Create a new value map model
-     */
+    Q_ENUM( ValueMapRoles )
+
     explicit ValueMapModel( QObject *parent = nullptr );
 
     /**
      * The map, see the property description
```
Comment (nirvn) [inline src/core/valuemapmodel.h:54]: Can you update these to rely on the proper `//! \copydoc ValueMapModel::valueMap`  for this function and the setMap function? We might as well take the opportunity to clean this up :)

File: src/core/valuemapmodelbase.h [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    Q_OBJECT
+
+    /**
+     * A list of QVariantMap, wrapped in a QVariant.
+     *
+     * Like this:
+     *
+     * [{'CH': 'Switzerland'}, {'DE': 'Germany'}, {'FR': 'France'}]
+     */
+    Q_PROPERTY( QVariant valueMap READ map WRITE setMap NOTIFY mapChanged )
```
Comment (nirvn) [inline src/core/valuemapmodelbase.h:39]: you can remove that, it's not going to be exposed in QML anyways.

File: src/core/valuemapmodelbase.h [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+    /**
+     * The map, see the property description
+     */
+
+    QVariant map() const;
+    /**
+     * The map, see the property description
+     */
+    void setMap( const QVariant &map );
```
Comment (nirvn) [inline src/core/valuemapmodelbase.h:55]: If you remove the Q_PROPERTY, you'll need to slightly tweak the description here.

File: src/qml/editorwidgets/ValueMap.qml (hunk 2 above)
Comment (domi4484) [reply src/qml/editorwidgets/ValueMap.qml:135]: Yes I'll look how far is feasible. I was scared about breaking something

File: src/core/valuemapmodelbase.h (hunk 5 above)
Comment (domi4484) [reply src/core/valuemapmodelbase.h:39]: Yes it is used here https://github.com/opengisch/QField/blob/3b03fff3432ed7b7dc618358f627dca404f00309/src/qml/editorwidgets/ValueMap.qml#L49C7-L49C12

Comment (domi4484) [reply src/core/valuemapmodelbase.h]: Oh I see now. In `AttributeFormModelBase` there where also some unused property. I removed them as well.

File: src/qml/editorwidgets/ValueMap.qml (hunk 2 above)
Comment (nirvn) [reply src/qml/editorwidgets/ValueMap.qml:135]: @domi4484 , shall we merge this PR now, and you can try to de-duplication code in a follow up PR?

Comment (nirvn) [reply src/qml/editorwidgets/ValueMap.qml:135]: @domi4484 , oh actually it'll need a rebase. PageHeader is no more, say hello to QfPageHeader :)

Comment (domi4484) [reply src/qml/editorwidgets/ValueMap.qml:135]: yes lets merge it

Comment (domi4484) [reply src/qml/editorwidgets/ValueMap.qml:135]: @nirvn Ill rebase!

Comment (domi4484) [reply src/qml/editorwidgets/ValueMap.qml:135]: @nirvn done

**Review by nirvn — APPROVED** [review]

**Pull request merged**